controllers/scopetemplate_controller.go

```diff
@@ -54,29 +71,58 @@ func (r *ScopeTemplateReconciler) Reconcile(ctx context.Context, req ctrl.Reques
 
 	log.Log.Info("Reconciling ScopeTemplate")
 
-	// get 	the scope template
-	in := &operatorsv1.ScopeTemplate{}
-	if err := r.Client.Get(ctx, req.NamespacedName, in); err != nil {
+	// get the scope template
+	st := &operatorsv1.ScopeTemplate{}
+	if err := r.Client.Get(ctx, req.NamespacedName, st); err != nil {
 		return ctrl.Result{}, err
 	}
 
-	// Create all the clusterRoles
-	for _, cr := range in.Spec.ClusterRoles {
-		clusterRole := &rbacv1.ClusterRole{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: cr.GenerateName,
-			},
-			Rules: cr.Rules,
-		}
+	scopeinstances := operatorsv1.ScopeInstanceList{}
 
-		log.Log.Info("Creating ClusterRole", "name", clusterRole.GenerateName)
-		if err := r.Client.Create(ctx, clusterRole); err != nil {
-			if !errors.IsAlreadyExists(err) {
-				return ctrl.Result{}, err
+	if err := r.Client.List(ctx, &scopeinstances, client.InNamespace(st.Namespace)); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	var flag bool
+	for i := range scopeinstances.Items {
+		sInstance := scopeinstances.Items[i]
+		if sInstance.Spec.ScopeTemplateName == st.Name {
+			// create ClusterRoles based on the ScopeTemplate
+			log.Log.Info("ScopeInstance found that references ScopeTemplate", "name", st.Name)
+			if err := r.ensureClusterRoles(ctx, st); err != nil {
+				return ctrl.Result{}, fmt.Errorf("Error in create ClusterRoles: %v", err)
 			}
+			flag = true
+		}
+	}
 
-			// TODO: Handle potential update.
+	if !flag {
+		listOption := client.MatchingLabels{
+			scopeTemplateUIDKey: string(st.GetUID()),
 		}
+
+		if err := r.deleteClusterRoles(ctx, listOption); err != nil {
+			log.Log.Info("Error in deleting Role Bindings", "error", err)
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	listOption := client.MatchingLabels{
+		scopeTemplateUIDKey: string(st.GetUID()),
+	}
+
+	requirement, err := labels.NewRequirement(scopeTemplateHashKey, selection.NotEquals, []string{util.HashObject(st.Spec)})
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	listOptions := &client.ListOptions{
+		LabelSelector: labels.NewSelector().Add(*requirement),
+	}
+
+	if err := r.deleteClusterRoles(ctx, listOption, listOptions); err != nil {
```

Comment:
Requires changes suggested above, could not incorporate all changes in one suggestion due to removed line.
```suggestion
	if err := r.deleteClusterRoles(ctx, listOptions...); err != nil {
```

Reply:
Updated.